### PR TITLE
CDAP-5380 Allow setting stream config/desc during creation of stream through REST API

### PR DESCRIPTION
**Pull request by gokulavasan**
JIRA :  https://issues.cask.co/browse/CDAP-5380 and https://issues.cask.co/browse/CDAP-5330

Build : http://builds.cask.co/browse/CDAP-DUT3943

### CODE REVIEW

**Comment by sreevatsanraman** [comment]
Can you change the docs and examples in the docs.

**Comment by gokulavasan** [comment]
@sreevatsanraman Updated. @awholegunch Can I take your help with editing the docs changes? I think it needs a bit of work in terms of linking to other section.

**Comment by awholegunch** [comment]
I've edited the docs. Please review my changes.

**Comment by awholegunch** [comment]
Should the default value of the stream description be `null`? If it is just a label, should it not just be an empty string ("")?

**Comment by gokulavasan** [comment]
@awholegunch I see that ``StreamSpecification`` has description as ``@Nullable``. So I followed the same convention in ``StreamProperties`` (these two classes need to be consolidated - will open a JIRA). @poornachandra Any thoughts? Also if you can review this PR that would be great!

**Comment by gokulavasan** [comment]
@chtyim Please review when you get a chance. Thanks!

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/CreateStreamCommand.java [hunk 1]
```diff
@@ -45,14 +57,28 @@ public CreateStreamCommand(StreamClient streamClient, CLIConfig cliConfig) {
   @Override
   public void perform(Arguments arguments, PrintStream output) throws Exception {
     String streamId = arguments.get(ArgumentName.NEW_STREAM.toString());
+    StreamProperties streamProperties = null;
 
-    streamClient.create(Id.Stream.from(cliConfig.getCurrentNamespace(), streamId));
+    if (arguments.hasArgument(ArgumentName.LOCAL_FILE_PATH.toString())) {
+      File file = new File(arguments.get(ArgumentName.LOCAL_FILE_PATH.toString()));
+      if (!file.isFile()) {
```
Comment (chtyim) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/CreateStreamCommand.java]: It's better not to check in here, but letting the creation of Reader to check and fail with an `IOException` below.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java [hunk 2]
```diff
@@ -194,9 +196,20 @@ public void create(HttpRequest request, HttpResponder responder,
       throw new BadRequestException(e);
     }
 
-    // TODO: Modify the REST API to support custom configurations.
-    streamAdmin.create(streamId);
+    StreamProperties properties = null;
+    // If the request to create a stream contains a non-empty body, then construct and set StreamProperties
+    if (request.getContent() != ChannelBuffers.EMPTY_BUFFER) {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java]: Don't use reference comparison. Just check `if request.getContent().readable()`

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java [hunk 3]
```diff
@@ -194,9 +196,20 @@ public void create(HttpRequest request, HttpResponder responder,
...
+      properties = getAndValidateConfig(request, responder);
+      if (properties == null) {
+        return;
+      }
+    }
 
+    // TODO: Consolidate the two separate calls to StreamAdmin
+    streamAdmin.create(streamId);
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java]: Why not just consolidate it? Is it a big change?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java [hunk 4]
```diff
@@ -472,7 +488,8 @@ public StreamProperties deserialize(JsonElement json, Type typeOfT,
       Integer threshold = jsonObj.has("notification.threshold.mb") ?
         jsonObj.get("notification.threshold.mb").getAsInt() :
         null;
-      return new StreamProperties(ttl, format, threshold);
+      String description = jsonObj.has("description") ? jsonObj.get("description").getAsString() : null;
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java]: Using `Optional` can have cleanup code:

```String desc = Optional.fromNullable(jsonObj.get("description")).orNull();```

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/StreamConfig.java [hunk 5]
```diff
@@ -41,18 +41,21 @@
...
   private final FormatSpecification format;
   private final int notificationThresholdMB;
+  private final String description;
 
   private final transient Location location;
 
   public StreamConfig(Id.Stream streamId, long partitionDuration, long indexInterval, long ttl,
-                      Location location, FormatSpecification format, int notificationThresholdMB) {
+                      Location location, FormatSpecification format, int notificationThresholdMB,
+                      String description) {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/StreamConfig.java]: Why store description in StreamConfig? Should it be just in MDS?

**Comment by gokulavasan** [comment]
@chtyim Addressed your comments. Please review again when you get a chance. Thanks!

**Comment by chtyim** [comment]
LGTM

**Pull request merged**